Require Comments for Status and Substatus #165919789

LoganArnett opened the pull request:
This PR covers [#165919789](https://www.pivotaltracker.com/story/show/165919789) for whether a status or substatus requires comments on the status modal and the validation on the form.

akegan commented:
@LoganArnett can you switch the base on this PR to master?

akegan commented:
@LoganArnett working on QAing the work that you've done so far now, but looks like this is missing AC 2+3, showing substatus in status history on the supp app page

akegan commented on app/javascript/components/organisms/StatusModalWrapper.js:
```diff
@@ -66,7 +68,7 @@ class StatusModalWrapper extends React.Component {
               </React.Fragment>
             )}
-            <label className='sr-only' htmlFor='status-comment' id='status-comment-label'>Comment</label>
+            <h2 className='form-label' id='status-comment-label'>Comment{statusRequiresComments(status, subStatus) ? ' (required)' : ''}</h2>
```
I think we can keep this as a label, and add a blocknote for required/not required, similar to how we do it in paper application form

LoganArnett replied:
reverted this back to a label

LoganArnett committed:
updating substatus in history and comment

LoganArnett committed:
updating substatus in history and comment

LoganArnett commented:
Updated both of these and resolved conflicts from master

LoganArnett committed:
updating unit tests

LoganArnett committed:
updating unit tests

LoganArnett committed:
updating tests for new comment structure in history

LoganArnett committed:
updating tests for new comment structure in history

akegan commented:
@LoganArnett looks like something went funny with the switch to master, looks like commits from react-final-form update are still on the branch (rather than just the squashed ones). If you interactively rebase onto master you should be able to just add your commits, but let me know if you have any trouble

akegan commented on spec/javascript/components/lease_ups/LeaseUpApplicationsPage.test.js:
```diff
@@ -160,24 +185,25 @@ describe('LeaseUpApplicationsPage', () => {
+        // Leave comment empty and submit
+        wrapper.find(updateButtonSelector).simulate('submit')
 
-      // Close the alert message
```
any particular reason why you've removed this test? I think we added it because of a bug we found where the red alert box wouldn't close

LoganArnett replied:
not sure how that was removed, I was using it as the basis of my other required test, just put it back in

akegan commented:
Sorry to keep piling on the comments, but I think that we want to only add error state to the fields with errors if that's possible

Status and Status detail are red even though there's nothing wrong with them:
![image](https://user-images.githubusercontent.com/11825994/58198227-b4a7cf80-7c82-11e9-8f20-b7efd07aa17a.png)

akegan commented:
@LoganArnett one last comment. Talked to emily and she thinks we should match the look of status update comments in history between top and bottom if that's possible! keeping the look of the ones at the bottom of the page now

LoganArnett committed:
replacing the failed submit test

LoganArnett committed:
replacing the failed submit test

LoganArnett commented:
For the errors comment I will look to see what can be done but it is currently setup to put the `error` class on the wrapper around everything in the form which causes the red on each label and the textarea

LoganArnett committed:
adjusting styling of substatus on top

LoganArnett committed:
adjusting styling of substatus on top

LoganArnett committed:
removing global form error and targeting comment directly

LoganArnett committed:
removing global form error and targeting comment directly

LoganArnett commented:
Updated the errors to only target the comment section which is the only thing truly required/validated currently and updated the substatus up top to reflect the history style

LoganArnett commented:
![image](https://user-images.githubusercontent.com/8780547/58203780-f186bd00-7ca8-11e9-9716-43c937e879e3.png)

akegan commented on spec/javascript/components/lease_ups/LeaseUpApplicationsPage.test.js:
```diff
@@ -179,5 +179,25 @@ describe('LeaseUpApplicationsPage', () => {
       expect(wrapper.find('.alert-body').exists()).toBe(false)
     })
+
+    test('Should open closeable alert modal on failed submit', async () => {
```
looks like you might have just wiped out your updated test. This one is a duplicate of above.

LoganArnett replied:
yep, replaced and pushed, I must have missed a commit in my rebase

LoganArnett committed:
replacing wiped out helper function

LoganArnett committed:
replacing wiped out helper function

LoganArnett committed:
updating tests

LoganArnett committed:
updating tests

LoganArnett committed:
updating for cases with and without substatus

LoganArnett committed:
updating for cases with and without substatus

akegan commented:
I'm seeing errors for comments being required even if (required) isn't on the comment box 
![image](https://user-images.githubusercontent.com/11825994/58209107-3efc2d80-7c9b-11e9-8fa2-3d988bb028f4.png)

akegan commented on app/javascript/components/organisms/StatusModalWrapper.js:
```diff
@@ -66,7 +67,7 @@ class StatusModalWrapper extends React.Component {
               </React.Fragment>
             )}
-            <label className='sr-only' htmlFor='status-comment' id='status-comment-label'>Comment</label>
+            <label className={`form-label ${formUtils.submitErrors(formApi).comment ? 'error' : ''}`} id='status-comment-label'>Comment{statusRequiresComments(status, subStatus) ? ' (required)' : ''}</label>
```
is it feasible to update this to add a [blocknote](https://github.com/Exygy/sf-dahlia-lap/blob/116d1351172dc1057acc68f2989a22c92615ae58/app/javascript/utils/form/final_form/Field.js#L10) or does that not make sense if we're not using react-final-form yet?

akegan commented:
I think I missed this in QA of previous PR, but "Status Detail" still shows up even if dropdown doesn't
![image](https://user-images.githubusercontent.com/11825994/58209440-f5601280-7c9b-11e9-82a5-b2e34ae3d2bf.png)

 We also need to require substatus/status detail when it's available.

LoganArnett committed:
updating removed test

LoganArnett committed:
updating removed test

LoganArnett committed:
revise status detail label

LoganArnett committed:
revise status detail label

LoganArnett commented:
The status detail label was also in whichever commit I missed earlier when rebasing, it has been updated

LoganArnett committed:
updating helper method for code climate

LoganArnett committed:
updating helper method for code climate

LoganArnett committed:
refactor test helpers for line limit

LoganArnett committed:
refactor test helpers for line limit

LoganArnett committed:
Merge branch 'master' of github.com:Exygy/sf-dahlia-lap into features/require-comments-status-substatus-165919789
updating with master

LoganArnett committed:
updating for comment error when not required

LoganArnett committed:
updating for comment error when not required

akegan commented:
@LoganArnett looks like substatus still needs to be required. If e.g. you update status to withdrawn and try to submit without adding a substatus you are blocked from submitting, but no error shows. 

Edit: If you then select a substatus that does not require a comment, no validation error message shows, but you're still blocked from submitting.

LoganArnett commented:
I began looking into this and found that it is a little bit of a rabbit hole to go down to get it working with react-form and seems like that would be time better spent refactoring to react-final-form instead of doing it twice

akegan commented:
ah ok, in that case are you up for taking on upgrading it to react-final-form? or maybe @software-project can take that over while we're all out on monday?

software-project commented:
@akegan sure, can do. I think this might be the thing we were talking about yesterday.

LoganArnett commented:
I'm up for it but I am limited until Wednesday so if it isnt a rush to get
completed I can

On Fri, May 24, 2019, 10:56 AM Marcin Jędras <notifications@github.com>
wrote:

> @akegan <https://github.com/akegan> sure, can do. I think this might be
> the thing we were talking about yesterday.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Exygy/sf-dahlia-lap/pull/249?email_source=notifications&email_token=ACC7WA6VTGKZ5KEYF7FKBX3PXAT5BA5CNFSM4HOFZHJKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWGEKOI#issuecomment-495732025>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACC7WA7SPUG2GCCGEFRQVDLPXAT5BANCNFSM4HOFZHJA>
> .
>

akegan commented:
ok @software-project let's see if you have time monday/tuesday to take it on

akegan commented:
@software-project maybe we can check in tomorrow about this PR and the remove react-final-form PR to figure out how to get both things done!

software-project commented:
@akegan yesterday I finished Remove remaining react-forms PR so the base for this one is ready.

software-project committed:
[#165919789] Updates for react final form

akegan commented:
@LoganArnett @software-project what's the status of this PR?

software-project commented:
@akegan @LoganArnett I rebased this PR of final form and I wasnt sure about the scope of change you wanted to make.

LoganArnett committed:
updating with remote

LoganArnett committed:
updating comment label id for testing

LoganArnett committed:
clean up and utilizing react-final-form state for performance instead of store

LoganArnett committed:
updating tests and test ids for the modal and labels

LoganArnett committed:
updating final-form fields for code-climate

LoganArnett committed:
updating e2e for final form substatus updates

akegan commented:
@LoganArnett is this ready for review again or still wip?

LoganArnett commented:
This still has a few updates to tests but for the functional part is ready for review @akegan

LoganArnett committed:
updating with master

LoganArnett commented:
@akegan I believe with updating the conflicts it should be good to go, I have all unit and e2e tests running locally

LoganArnett commented:
I am unable to get these e2e tests to fail locally, I am going to SSH into the instance to see if there is anything else that can help me track the errors down but I am uncertain at this point

akegan commented:
@LoganArnett which test is having issues on semaphore? Is it possible that it's being flaky based on some state of the application?

LoganArnett commented:
@akegan it is both tests in `spec/javascript/e2e/SupplementalApplicationPage/status_update.e2e.js` and they are consistently failing on semaphore but consistently passing locally

akegan reviewed (changes requested):
Just did some QA and found a few more issues:

1. add block note (required) to status detail if it's required
1. if status detail is required and not present, highlight dropdown in red and change field label to red too
1. For lease signed, comments should not be required but user is blocked from saving without comment
1. For Withdrawal: For written withdrawal, comments are not required and behavior looks good. For verbal withdrawal, comments should be required but user can save without entering comment.

LoganArnett commented:
There were a few regressions from the final-form switch and I thought I caught them all but apparently not, I will update these ASAP

akegan commented on app/javascript/utils/form/final_form/Field.js:
```diff
@@ -133,7 +132,18 @@ export const CheckboxField = ({ fieldName, label, blockNote, validation, id, ari
 )
 
-export const YesNoRadioField = ({ fieldName, uniqId, trueValue = 'true', trueLabel = 'Yes', falseValue = 'false', falseLabel = 'No', inputClassName, className, label }) => {
+const YesNoRadioField = ({ value, label, type, fieldName, className, uniqId }) => (
```
curious where these updates to YesNoRadio field/group came from in this PR. if just a refactor or potential regression

LoganArnett replied:
These were due to code climate failing on similar code issues, refactored to meet the standard

akegan commented on spec/javascript/components/supplemental_application/sections/__snapshots__/RentalAssistance.test.js.snap:
```diff
@@ -566,8 +573,9 @@ Array [
+              id="label-recipient"
             >
-              Recipient 
+              Recipient
```
also curious where these removed whitespaces come from

LoganArnett replied:
not sure why whitespace would be removed for the snapshots

akegan reviewed (commented):
couple more questions

LoganArnett committed:
testing validate on blur issue

LoganArnett committed:
updating based on form level validation and not on blur

LoganArnett committed:
updating status detail always required and verbal issue

LoganArnett committed:
Merge branch 'master' of github.com:Exygy/sf-dahlia-lap into features/require-comments-status-substatus-165919789
updating with master

LoganArnett committed:
updating for the lease signed logic with explicit can submit logic

LoganArnett committed:
refactor if statement

LoganArnett committed:
updating for required issue

LoganArnett committed:
updating with master and snapshot

LoganArnett commented:
So Semaphore is successful now but I did have to rebuild several times without changes to get past false negatives. Now code climate is throwing a duplicate code error on the `react-final-form` components we are using but the 2 locations it is mentioning do not appear to have any duplicate code other than the `Form` that we have to use from `react-final-form`.

LoganArnett commented:
Also found that code climate has had and continues to have issues with similar setups to ours when trying to create reusable components: https://github.com/codeclimate/codeclimate-duplication/issues/326

akegan commented:
ok @LoganArnett no worries. take code climate duplication/refactoring issues as suggestions rather than hard and fast requirements. We're more interested in the test coverage aspect

chrisdolendo commented:
@akegan @LoganArnett semaphore passes now - if this is approved we can merge

LoganArnett committed:
Merge branch 'master' into features/require-comments-status-substatus-165919789

akegan commented on app/javascript/components/applications/application_form/DemographicInfoSection.js:
```diff
@@ -61,8 +54,7 @@ const DemographicInfoSection = ({ values: { demographics } }) => {
             label='Sexual Orientation (if not listed)'
-            maxLength={maxLengthMap['sexual_orientation_other']}
-            validation={validateNotListed('Sexual Orientation is required', 'sexual_orientation')} />
+            maxLength={maxLengthMap['sexual_orientation_other']} />
```
@LoganArnett looks like some sort of merge conflict here undoing your gender validation changes?

LoganArnett replied:
This is actually in relation to how the validation was working and how we are not using the direct react-final-form values and handlers connected so I put the validation at the form level instead of the field level so that the tests pass on blur or change

akegan replied:
ok I will re-test this stuff

LoganArnett committed:
Merge branch 'master' of github.com:Exygy/sf-dahlia-lap into features/require-comments-status-substatus-165919789
updating with master

LoganArnett committed:
Merge branch 'features/require-comments-status-substatus-165919789' of github.com:Exygy/sf-dahlia-lap into features/require-comments-status-substatus-165919789
updating with remote

akegan commented:
looks good! go ahead and merge!
